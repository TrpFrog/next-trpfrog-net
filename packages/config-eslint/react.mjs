// @ts-check
import { createESLintConfig } from './base.mjs'
import pluginReact from 'eslint-plugin-react'

/**
 * ESLint config for React projects.
 * @param {import('typescript-eslint').ConfigWithExtends[]} userConfig
 * @returns
 */
export const createESLintConfigReact = (...userConfig) =>
  createESLintConfig({
    files: ['**/*.tsx'],
    plugins: {
      react: pluginReact,
    },
  })
