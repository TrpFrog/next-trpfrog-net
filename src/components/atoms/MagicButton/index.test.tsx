import { render, screen } from '@testing-library/react'

import { MagicButton } from '.'

describe('Button', () => {
  describe('div mode', () => {
    test('should render correctly', () => {
      render(<MagicButton>test</MagicButton>)
      expect(screen.getByTestId('button-component').tagName).toBe('DIV')
    })

    test('snapshot testing', () => {
      const { asFragment } = render(<MagicButton>test</MagicButton>)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('button mode', () => {
    test('should render correctly', () => {
      render(<MagicButton onClick={() => {}}>test</MagicButton>)
      expect(screen.getByTestId('button-component').tagName).toBe('BUTTON')
    })

    test('snapshot testing', () => {
      const { asFragment } = render(
        <MagicButton onClick={() => {}}>test</MagicButton>,
      )
      expect(asFragment()).toMatchSnapshot()
    })

    test('should call onClick', () => {
      const onClick = vi.fn()
      render(<MagicButton onClick={onClick}>test</MagicButton>)
      screen.getByTestId('button-component').click()
      expect(onClick).toBeCalledTimes(1)
    })

    test('should not call onClick when disabled', () => {
      const onClick = vi.fn()
      render(
        <MagicButton onClick={onClick} disabled>
          test
        </MagicButton>,
      )
      screen.getByTestId('button-component').click()
      expect(onClick).toBeCalledTimes(0)
    })
  })

  describe('a mode', () => {
    test('should render correctly', () => {
      render(<MagicButton href="/">test</MagicButton>)
      expect(screen.getByTestId('button-component').tagName).toBe('A')
    })

    test('snapshot testing (internal link)', () => {
      const { asFragment } = render(<MagicButton href="/">test</MagicButton>)
      expect(asFragment()).toMatchSnapshot()
    })

    test('snapshot testing (external link)', () => {
      const { asFragment } = render(
        <MagicButton href="https://github.com">test</MagicButton>,
      )
      expect(asFragment()).toMatchSnapshot()
    })

    const internalLinks = [
      ['/'],
      ['/test'],
      ['#'],
      ['mailto:test@example'],
      ['https://trpfrog.net'],
    ]
    test.each(internalLinks)(
      'should open internal link in current tab (%s)',
      href => {
        render(<MagicButton href={href}>test</MagicButton>)
        expect(screen.getByTestId('button-component')).not.toHaveAttribute(
          'target',
          '_blank',
        )
      },
    )
    test.each(internalLinks)(
      'should open internal link in new tab if there is an externalLink attr (%s)',
      href => {
        render(
          <MagicButton externalLink href={href}>
            test
          </MagicButton>,
        )
        expect(screen.getByTestId('button-component')).toHaveAttribute(
          'target',
          '_blank',
        )
      },
    )

    const externalLinks = [
      ['https://www.apple.com'],
      ['https://www.google.com'],
      ['https://github.com'],
    ]
    test.each(externalLinks)(
      'should open external link in new tab (%s)',
      href => {
        render(<MagicButton href={href}>test</MagicButton>)
        expect(screen.getByTestId('button-component')).toHaveAttribute(
          'target',
          '_blank',
        )
      },
    )
  })

  describe('disabled', () => {
    test('should render correctly', () => {
      render(<MagicButton disabled>test</MagicButton>)
      expect(screen.getByTestId('button-component')).toHaveAttribute('disabled')
    })

    test('should not call onClick', () => {
      const onClick = vi.fn()
      render(
        <MagicButton onClick={onClick} disabled>
          test
        </MagicButton>,
      )
      screen.getByTestId('button-component').click()
      expect(onClick).not.toHaveBeenCalled()
    })

    test('should not jump to /', () => {
      render(
        <MagicButton href="/" disabled>
          test
        </MagicButton>,
      )
      expect(screen.getByTestId('button-component')).toHaveAttribute(
        'href',
        '/',
      )
    })
  })
})
