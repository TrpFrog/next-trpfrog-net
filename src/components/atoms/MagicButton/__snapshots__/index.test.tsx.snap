// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button a mode snapshot testing (external link) 1`] = `
<DocumentFragment>
  <a
    class="button"
    data-testid="button-component"
    href="https://github.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    test
  </a>
</DocumentFragment>
`;

exports[`Button a mode snapshot testing (internal link) 1`] = `
<DocumentFragment>
  <a
    class="button"
    data-testid="button-component"
    href="/"
  >
    test
  </a>
</DocumentFragment>
`;

exports[`Button button mode snapshot testing 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="button-component"
  >
    test
  </button>
</DocumentFragment>
`;

exports[`Button div mode snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="button"
    data-testid="button-component"
  >
    test
  </div>
</DocumentFragment>
`;
